src/datasets/arrow_writer.py

```diff
@@ -99,6 +100,11 @@ def __arrow_array__(self, type=None):
                 raise TypeError(
                     "Specified try_type alters data. Please check that the type/feature that you provided match the type/features of the data."
                 )
+            if self.optimized_int_type and self.type is None and self.try_type is None:
+                if pa.types.is_int64(out.type):
+                    out = out.cast(self.optimized_int_type)
+                elif pa.types.is_list(out.type) and pa.types.is_int64(out.type.value_type):
+                    out = out.cast(pa.list_(self.optimized_int_type))
```

Comment:
@lhoestq might there be other cases besides scalar int and list of int?

Reply:
Maybe add list of list of int but honestly this is an extremely rare case

Reply:
Done.